src/proof/verify.rs

```diff
@@ -17,71 +19,100 @@ use nybbles::Nibbles;
 pub fn verify_proof<'a, I>(
     root: B256,
     key: Nibbles,
-    value: Option<Vec<u8>>,
+    expected_value: Option<Vec<u8>>,
     proof: I,
 ) -> Result<(), ProofVerificationError>
 where
     I: IntoIterator<Item = &'a Bytes>,
 {
     let mut proof = proof.into_iter().peekable();
 
+    // If the proof is empty or contains only an empty node, the expected value must be None.
     if proof.peek().map_or(true, |node| node.as_ref() == [EMPTY_STRING_CODE]) {
         return if root == EMPTY_ROOT_HASH {
-            if value.is_none() {
+            if expected_value.is_none() {
                 Ok(())
             } else {
                 Err(ProofVerificationError::ValueMismatch {
                     path: key,
                     got: None,
-                    expected: value.map(Bytes::from),
+                    expected: expected_value.map(Bytes::from),
                 })
             }
         } else {
             Err(ProofVerificationError::RootMismatch { got: EMPTY_ROOT_HASH, expected: root })
         };
     }
 
-    let mut walked_path = Nibbles::default();
-    let mut next_value = Some(RlpNode::word_rlp(&root));
+    let mut walked_path = Nibbles::with_capacity(key.len());
+    let mut last_decoded_node = Some(NodeDecodingResult::Node(RlpNode::word_rlp(&root)));
```

Comment:
can be a smallvec, no need for enum then. seems a bit too verbose

Reply:
It would be more verbose with smallvec, I don't mind this

Reply:
it would be hard to do without copying, given that `RlpNode` isn't always 33 bytes so we can't use `into_inner` to obtain a const array, and we will need to create a `SmallVec` from a slice.